File: openvidu-call/openvidu-call-back/src/controllers/SessionController.ts
import * as express from 'express';
import * as crypto from 'crypto';
import { Request, Response } from 'express';
import { OpenViduRole, Session } from 'openvidu-node-client';
import { OpenViduService } from '../services/OpenViduService';
import { RECORDING } from '../config';

export const app = express.Router({
	strict: true
});

const openviduService = OpenViduService.getInstance();

app.post('/', async (req: Request, res: Response) => {
	try {
		console.log('Session ID received', req.body.sessionId);

		let sessionId: string = req.body.sessionId;
		let nickname: string = req.body.nickname;
		let date = null;
		let sessionCreated: Session = await openviduService.createSession(sessionId);
		const RECORDING_TOKEN_NAME = openviduService.RECORDING_TOKEN_NAME;
		const IS_RECORDING_ENABLED = RECORDING.toUpperCase() === 'ENABLED';
		const hasValidToken = openviduService.isValidToken(sessionId, req.cookies);
		const isSessionCreator = hasValidToken || sessionCreated.activeConnections.length === 0;
		const role: OpenViduRole = isSessionCreator && IS_RECORDING_ENABLED ? OpenViduRole.MODERATOR : OpenViduRole.PUBLISHER;
		const response = {cameraToken : '', screenToken: '', recordingEnabled: IS_RECORDING_ENABLED, recordings: []};
		const cameraConnection = await openviduService.createConnection(sessionCreated, nickname, role);
		const screenConnection = await openviduService.createConnection(sessionCreated, nickname, role);
		response.cameraToken = cameraConnection.token;
		response.screenToken = screenConnection.token;

		if (IS_RECORDING_ENABLED && isSessionCreator && !hasValidToken) {
			/**
			 * ! *********** WARN *********** !
			 *
			 * To identify who is able to manage session recording, the code sends a cookie with a token to the session creator.
			 * The relation between cookies and sessions are stored in backend memory.
			 *
			 * This authentication & authorization system is pretty basic and it is not for production.
			 * We highly recommend IMPLEMENT YOUR OWN USER MANAGEMENT with persistence for a properly and secure recording feature.
			 *
			 * ! *********** WARN *********** !
			 **/
			const uuid = crypto.randomBytes(32).toString('hex');
			date = Date.now();
			const recordingToken = `${response.cameraToken}&${RECORDING_TOKEN_NAME}=${uuid}&createdAt=${date}`;
			res.cookie(RECORDING_TOKEN_NAME, recordingToken);
			openviduService.recordingMap.set(sessionId, { token: recordingToken, recordingId: '' });
		}

		if(IS_RECORDING_ENABLED){
			date = date || openviduService.getDateFromCookie(req.cookies);
			try {
				response.recordings = await openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
			} catch (error) {
				if(error.message === '501'){
					console.log('Recording is diasbled in OpenVidu Server. Disabling it in OpenVidu Call');
					response.recordings = [];
					response.recordingEnabled = false;
				}
			}
		}

		res.status(200).send(JSON.stringify(response));
	} catch (error) {
		console.error(error);
		let message = 'Cannot connect with OpenVidu Server';
		let code = Number(error?.message);
		if(error.message === 500){
			message = 'Unexpected error when creating the Connection object.'
		} else if (error.message === 404){
			message = 'No session exists';
		}
		if(typeof code !== 'number') {
			code = 503
		}
		res.status(code).send({ message });
	}
});

File: openvidu-call/openvidu-call-back-java/src/main/java/io/openvidu/call/java/controllers/SessionController.java
package io.openvidu.call.java.controllers;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import io.openvidu.call.java.models.RecordingData;
import io.openvidu.call.java.services.OpenViduService;
import io.openvidu.java.client.Connection;
import io.openvidu.java.client.OpenViduHttpException;
import io.openvidu.java.client.OpenViduJavaClientException;
import io.openvidu.java.client.OpenViduRole;
import io.openvidu.java.client.Recording;
import io.openvidu.java.client.Session;

@CrossOrigin(origins = "*")
@RestController
public class SessionController {

	@Value("${RECORDING}")
	private String RECORDING;

	@Autowired
	private OpenViduService openviduService;

	@PostMapping("/sessions")
	public ResponseEntity<Map<String, Object>> createConnection(
			@RequestBody(required = false) Map<String, Object> params,
			@CookieValue(name = OpenViduService.RECORDING_TOKEN_NAME, defaultValue = "") String recordingTokenCookie,
			HttpServletResponse res) {

		Map<String, Object> response = new HashMap<String, Object>();
		try {
			long date = -1;
			String nickname = "";

			String sessionId = params.get("sessionId").toString();
			if (params.containsKey("nickname")) {
				nickname = params.get("nickname").toString();
			}

			Session sessionCreated = this.openviduService.createSession(sessionId);
			boolean IS_RECORDING_ENABLED = RECORDING.toUpperCase().equals("ENABLED");

			boolean hasValidToken = this.openviduService.isValidToken(sessionId, recordingTokenCookie);
			boolean isSessionCreator = hasValidToken || sessionCreated.getActiveConnections().size() == 0;

			OpenViduRole role = isSessionCreator && IS_RECORDING_ENABLED ? OpenViduRole.MODERATOR
					: OpenViduRole.PUBLISHER;

			response.put("recordingEnabled", IS_RECORDING_ENABLED);
			response.put("recordings", new ArrayList<Recording>());

			Connection cameraConnection = this.openviduService.createConnection(sessionCreated, nickname, role);
			Connection screenConnection = this.openviduService.createConnection(sessionCreated, nickname, role);

			response.put("cameraToken", cameraConnection.getToken());
			response.put("screenToken", screenConnection.getToken());

			if (IS_RECORDING_ENABLED && isSessionCreator && !hasValidToken) {
				/**
				 * ! *********** WARN *********** !
				 *
				 * To identify who is able to manage session recording, the code sends a cookie
				 * with a token to the session creator. The relation between cookies and
				 * sessions are stored in backend memory.
				 *
				 * This authentication & authorization system is pretty basic and it is not for
				 * production. We highly recommend IMPLEMENT YOUR OWN USER MANAGEMENT with
				 * persistence for a properly and secure recording feature.
				 *
				 * ! *********** WARN *********** !
				 **/

				String uuid = UUID.randomUUID().toString();
				date = System.currentTimeMillis();
				String recordingToken = cameraConnection.getToken() + "&" + OpenViduService.RECORDING_TOKEN_NAME + "="
						+ uuid + "&createdAt=" + date;

				Cookie cookie = new Cookie(OpenViduService.RECORDING_TOKEN_NAME, recordingToken);
				res.addCookie(cookie);

				RecordingData recData = new RecordingData(recordingToken, "");
				this.openviduService.recordingMap.put(sessionId, recData);
			}

			if (IS_RECORDING_ENABLED) {
				if (date == -1) {
					date = openviduService.getDateFromCookie(recordingTokenCookie);
				}
				List<Recording> recordings = openviduService.listRecordingsBySessionIdAndDate(sessionId, date);
				response.put("recordings", recordings);
			}

			return new ResponseEntity<>(response, HttpStatus.OK);

		} catch (OpenViduJavaClientException | OpenViduHttpException e) {

			if (e.getMessage() != null && Integer.parseInt(e.getMessage()) == 501) {
				System.err.println("OpenVidu Server recording module is disabled");
				response.put("recordingEnabled", false);
				return new ResponseEntity<>(response, HttpStatus.OK);
			} else if (e.getMessage() != null && Integer.parseInt(e.getMessage()) == 401) {
				System.err.println("OpenVidu credentials are wrong.");
				return new ResponseEntity<>(null, HttpStatus.UNAUTHORIZED);

			} else {
				e.printStackTrace();
				System.err.println(e.getMessage());
				return new ResponseEntity<>(response, HttpStatus.INTERNAL_SERVER_ERROR);
			}
		}
	}

}
